pkg/apis/pipeline/v1beta1/pipeline_validation_test.go

```diff
@@ -315,26 +378,37 @@ func TestPipelineSpec_Validate_Failure(t *testing.T) {
 				},
 			}},
 		},
-	}, {
-		name: "invalid pipeline spec with DAG having cyclic dependency",
-		ps: &PipelineSpec{
-			Tasks: []PipelineTask{{
-				Name: "foo", TaskRef: &TaskRef{Name: "foo-task"}, RunAfter: []string{"bar"},
-			}, {
-				Name: "bar", TaskRef: &TaskRef{Name: "bar-task"}, RunAfter: []string{"foo"},
-			}},
-		},
+		expectedError: *apis.ErrGeneric(`invalid value: couldn't add link between foo and bar: task foo depends on bar but bar wasn't present in Pipeline`, "tasks").Also(
+			apis.ErrInvalidValue("expected resource great-resource to be from task bar, but task bar doesn't exist", "tasks[1].resources.inputs[0].from")),
 	}}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			err := tt.ps.Validate(context.Background())
 			if err == nil {
 				t.Errorf("PipelineSpec.Validate() did not return error for invalid pipelineSpec: %s", tt.name)
 			}
+			if d := cmp.Diff(tt.expectedError.Error(), err.Error(), cmpopts.IgnoreUnexported(apis.FieldError{})); d != "" {
+				t.Errorf("PipelineSpec.Validate() errors diff %s", diff.PrintWantGot(d))
+			}
 		})
 	}
 }
 
+func TestPipelineSpec_Validate_Failure_CycleDAG(t *testing.T) {
+	name := "invalid pipeline spec with DAG having cyclic dependency"
+	ps := &PipelineSpec{
+		Tasks: []PipelineTask{{
+			Name: "foo", TaskRef: &TaskRef{Name: "foo-task"}, RunAfter: []string{"bar"},
+		}, {
+			Name: "bar", TaskRef: &TaskRef{Name: "bar-task"}, RunAfter: []string{"foo"},
+		}},
+	}
+	err := ps.Validate(context.Background())
+	if err == nil {
+		t.Errorf("PipelineSpec.Validate() did not return error for invalid pipelineSpec: %s", name)
+	}
+}
+
```

Comment:
Extracted this to not have a flaky unit test. We do validate it errors out but not checking the error as the order is not deterministic (can be `foo -> bar -> foo` or `bar -> foo -> bar`, both make senses).